internal/engine/compiler/impl_amd64.go

```diff
@@ -269,13 +269,30 @@ func (c *amd64Compiler) label(label wazeroir.Label) *amd64LabelInfo {
 
 // compileBuiltinFunctionCheckExitCode implements compiler.compileBuiltinFunctionCheckExitCode for the amd64 architecture.
 func (c *amd64Compiler) compileBuiltinFunctionCheckExitCode() error {
-	if err := c.compileCallBuiltinFunction(builtinFunctionIndexCheckExitCode); err != nil {
+	dstReg, err := c.allocateRegister(registerTypeGeneralPurpose)
+	if err != nil {
+		panic("BUG: cannot find free register")
+	}
+
+	c.assembler.CompileMemoryToRegister(amd64.MOVQ,
+		amd64ReservedRegisterForCallEngine, callEngineModuleContextModuleInstanceOffset, dstReg)
+	c.assembler.CompileMemoryToRegister(amd64.MOVQ, dstReg, 0, dstReg)
+
+	// Compare the conditional value with zero.
+	c.assembler.CompileRegisterToConst(amd64.CMPQ, dstReg, 0)
+
+	jump := c.assembler.CompileJump(amd64.JEQ)
+
+	if err := c.compileCallBuiltinFunction(builtinFunctionIndexExitUnconditionally); err != nil {
 		return err
 	}
 
 	// After the function call, we have to initialize the stack base pointer and memory reserved registers.
 	c.compileReservedStackBasePointerInitialization()
 	c.compileReservedMemoryPointerInitialization()
```

Comment:
these lines can be deleted (same goes for arm64 as well) since we don't exit the machine code world anymore in this branch